File: projects/cahoots-eclipse/src/main/scala/com/cahoot/eclipse/collab/DisconnectDialog.scala
package com.cahoot.eclipse.collab

import scala.collection.mutable.ListBuffer
import org.apache.commons.httpclient.HttpClient
import org.apache.commons.httpclient.NameValuePair
import org.apache.commons.httpclient.methods.PostMethod
import org.eclipse.jface.dialogs.MessageDialog
import org.eclipse.swt.SWT
import org.eclipse.swt.widgets.Button
import org.eclipse.swt.widgets.Composite
import org.eclipse.swt.widgets.Label
import com.cahoot.eclipse.core.PluginConstants
import com.cahoot.eclipse.widget.CahootMigOkCancelDialog
import com.cahoot.eclipse.widget.listener.SwtListeners
import com.cahoot.models.PersonModel
import com.cahoot.eclipse.Activator
import net.miginfocom.swt.MigLayout
import com.cahoot.eclipse.ActiveUser
import com.cahoot.eclipse.ActiveUser

/**
 * Modal dialog for selecting which collaborators the user wishes to share with
 */
class DisconnectDialog() extends CahootMigOkCancelDialog {

  var help: Label = _
  var manage: Button = _

  val collaborators = new ListBuffer[PersonModel]

  override def setupContents(panel: Composite): Unit = {

    panel.setLayout(new MigLayout("fill", "",
      "[growprio 0][growprio 0][growprio 100][growprio 0]"))
      
    help = new Label(panel, SWT.NONE)
    help.setText("Are you sure you would like to disconnect from the server?")
    help.setLayoutData("")
  }

  override def setupDefaults(panel: Composite): Unit = {
    val shell = getShell()
    shell.setText("%s Collaborator".format(PluginConstants.APP_TITLE))
    shell.setSize(800, 600)
  }

  override def addListeners(): Unit = {

    ok.addSelectionListener(SwtListeners.selectionListener({ e =>
      var server: String = "http://localhost:9000";
      
      var client: HttpClient = new HttpClient();
      var method: PostMethod = new PostMethod( server + "/app/logout");
      
      var  data: Array[NameValuePair] = Array(
          new NameValuePair("auth_token", ActiveUser.auth_token)
        )
      method.setRequestBody(data)
      
      var statusCode: Int = client.executeMethod(method)
      
      if(statusCode == 200)
      {
    	ActiveUser.auth_token = ""
        var token:String = new String(method.getResponseBody())
        
      	close
      } 
      else 
      {
        //TODO better message on error
        
	    MessageDialog.openInformation(
				null,
				"Disconnect",
				"Error disconnecting from the server");
      }
      
    }))

    cancel.addSelectionListener(SwtListeners.selectionListener({ e =>
      close
    }))
  }

  def getCollaborators(): List[PersonModel] = collaborators.toList
  

}

File: projects/cahoots-eclipse/src/main/java/com/cahoots/eclipse/collab/DisconnectDialog.java
package com.cahoots.eclipse.collab;

import java.io.IOException;
import java.util.LinkedList;
import java.util.List;

import org.apache.commons.httpclient.HttpClient;
import org.apache.commons.httpclient.HttpException;
import org.apache.commons.httpclient.NameValuePair;
import org.apache.commons.httpclient.methods.PostMethod;
import org.eclipse.jface.dialogs.MessageDialog;
import org.eclipse.jface.window.Window;
import org.eclipse.swt.SWT;
import org.eclipse.swt.events.SelectionAdapter;
import org.eclipse.swt.events.SelectionEvent;
import org.eclipse.swt.widgets.Button;
import org.eclipse.swt.widgets.Composite;
import org.eclipse.swt.widgets.Label;
import org.eclipse.swt.widgets.Shell;

import com.cahoots.connection.ConnectionDetails;
import com.cahoots.connection.websocket.CahootsRealtimeClient;
import com.cahoots.eclipse.Activator;
import com.google.inject.Injector;

public class DisconnectDialog extends Window {

	private ConnectionDetails ConnectionDetails;
	private CahootsRealtimeClient cahootsSocket;

	/**
	 * Create the dialog.
	 * 
	 * @param parent
	 * @param style
	 */
	public DisconnectDialog(final Shell parent) {
		super(parent);

		setShellStyle((getShellStyle() | SWT.DIALOG_TRIM | SWT.APPLICATION_MODAL)
				& ~(SWT.RESIZE | SWT.MAX));
		final Injector injector = Activator.getInjector();
		ConnectionDetails = injector.getInstance(ConnectionDetails.class);
		cahootsSocket = injector.getInstance(CahootsRealtimeClient.class);
	}

	/**
	 * Create contents of the dialog.
	 */
	@Override
	protected Composite createContents(final Composite parent) {
		final Shell shell = getShell();
		shell.setSize(262, 91);
		shell.setText("Cahoots - Disconnect");
		shell.setLayout(null);

		final Composite content = parent;
		// Disconnect label
		final Label disconnectLabel = new Label(content, SWT.NONE);
		disconnectLabel.setBounds(10, 10, 244, 15);
		disconnectLabel.setText("Would you like to disconnect from Cahoots?");

		// Disconnect button
		final Button disconnect = new Button(content, SWT.NONE);
		disconnect.setBounds(98, 31, 75, 25);
		disconnect.setText("Disconnect");
		disconnect.addSelectionListener(new SelectionAdapter() {
			@Override
			public void widgetSelected(final SelectionEvent e) {
				final HttpClient client = new HttpClient();
				final PostMethod method = new PostMethod("http://"
						+ ConnectionDetails.getServer() + "/app/logout");
				final List<NameValuePair> data = new LinkedList<NameValuePair>();
				data.add(new NameValuePair("auth_token", ConnectionDetails
						.getAuthToken()));

				method.setRequestBody(data.toArray(new NameValuePair[data
						.size()]));
				try {
					client.executeMethod(method);
					performDisconnect();
				} catch (final HttpException ex) {
					MessageDialog.openInformation(
							null,
							"Disconnect Error",
							"Error disconnecting from server. "
									+ ex.getMessage());
				} catch (final IOException ex) {
					MessageDialog.openInformation(
							null,
							"Disconnect Error",
							"Error disconnecting from server. "
									+ ex.getMessage());
				}
			}

			private void performDisconnect() {
				ConnectionDetails.disconnect();
				cahootsSocket.disconnect();
				close();
			}
		});

		// Cancel button
		final Button cancel = new Button(content, SWT.NONE);
		cancel.setBounds(179, 31, 75, 25);
		cancel.setText("Cancel");
		cancel.addSelectionListener(new SelectionAdapter() {
			@Override
			public void widgetSelected(final SelectionEvent e) {
				shell.close();
			}
		});
		return content;
	}
}
